From 876ea76879d60e03920e0a8d47b76d13ee82ea2f Mon Sep 17 00:00:00 2001
From: Juergen Repp <juergen_repp@web.de>
Date: Fri, 22 Aug 2025 15:36:02 +0200
Subject: [PATCH 22/22] FAPI: Fix nv object authorization for policy authorize
 nv.

The cleanup function for the nv object to be authorized was called before
authorization, and the wrong object was used for authorization.

Signed-off-by: Juergen Repp <juergen_repp@web.de>
---
 src/tss2-fapi/ifapi_policy_callbacks.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/src/tss2-fapi/ifapi_policy_callbacks.c b/src/tss2-fapi/ifapi_policy_callbacks.c
index 9c52088d..99748176 100644
--- a/src/tss2-fapi/ifapi_policy_callbacks.c
+++ b/src/tss2-fapi/ifapi_policy_callbacks.c
@@ -1570,7 +1570,6 @@ ifapi_exec_auth_nv_policy(
             goto_if_error(r, "Initialize NV object", cleanup);
 
             current_policy->nv_index = cb_ctx->object.public.handle;
-            ifapi_cleanup_ifapi_object(&cb_ctx->object);
             get_nv_auth_object(&cb_ctx->object,
                                current_policy->nv_index,
                                &current_policy->auth_objectNV,
@@ -1579,8 +1578,10 @@ ifapi_exec_auth_nv_policy(
 
         statecase(cb_ctx->cb_state, POL_CB_AUTHORIZE_OBJECT)
             /* Authorize the NV object with the corresponding auth object. */
-            r = ifapi_authorize_object(fapi_ctx, &cb_ctx->auth_object, &cb_ctx->session);
+            r = ifapi_authorize_object(fapi_ctx, &current_policy->auth_objectNV, &cb_ctx->session);
             return_try_again(r);
+
+            ifapi_cleanup_ifapi_object(&cb_ctx->object);
             goto_if_error(r, "Authorize  object.", cleanup);
 
             /* Prepare the reading of the NV index from TPM. */
-- 
2.51.0

